tests/unit/model/relationships/rollback-test.js

```diff
@@ -0,0 +1,275 @@
+    });
+
+    env = setupStore({ person: Person, dog: Dog });
+    store = env.store;
+  }
+});
+
+test("saved changes to relationships should not roll back to a pre-saved state (from child)", function(assert) {
+  var person1, person2, dog1, dog2, dog3;
+
+  env.adapter.updateRecord = function(store, type, snapshot) {
+    return Ember.RSVP.resolve({ id: 2, owner: 1 });
+  };
+
+  run(function() {
+    store.push({
+      data: {
+        type: 'person',
+        id: 1,
+        attributes: {
+          firstName: "Tom",
+          lastName: "Dale"
+        },
+        relationships: {
+          dogs: {
+            data: [{
+              type: 'dog',
+              id: 1
+            }]
+          }
+        }
+      }
+    });
+    store.push({
+      data: {
+        type: 'person',
+        id: 2,
+        attributes: {
+          firstName: "John",
+          lastName: "Doe"
+        },
+        relationships: {
+          dogs: {
+            data: [{
+              type: 'dog',
+              id: 2
+            }]
+          }
+        }
+      }
+    });
+    store.push({
+      data: {
+        type: 'dog',
+        id: 1,
+        attributes: {
+          name: "Fido"
+        },
+        relationships: {
+          owner: {
+            data: {
+              type: 'person',
+              id: 1
+            }
+          }
+        }
+      }
+    });
+    store.push({
+      data: {
+        type: 'dog',
+        id: 2,
+        attributes: {
+          name: "Bear"
+        },
+        relationships: {
+          owner: {
+            data: {
+              type: 'person',
+              id: 2
+            }
+          }
+        }
+      }
+    });
+    store.push({
+      data: {
+        type: 'dog',
+        id: 3,
+        attributes: {
+          name: "Spot"
+        },
+        relationships: {
+          owner: {
+            data: null
+          }
+        }
+      }
+    });
+    person1 = store.peekRecord('person', 1);
+    person2 = store.peekRecord('person', 2);
+    dog1 = store.peekRecord('dog', 1);
+    dog2 = store.peekRecord('dog', 2);
+    dog3 = store.peekRecord('dog', 3);
+
+    person1.get('dogs').addObject(dog2);
+  });
+
+  run(function() {
+    dog2.save().then(function () {
+      person1.get('dogs').addObject(dog3);
+      return Ember.RSVP.all([dog2.rollback(), dog3.rollback()]);
+    }).then(function () {
+      person1.get('dogs').then(function (dogs) {
+        assert.deepEqual(dogs.toArray(), [dog1,dog2]);
+      });
+      person2.get('dogs').then(function (dogs) {
+        assert.deepEqual(dogs.toArray(), []);
+      });
+      dog1.get('owner').then(function (owner) {
+        assert.equal(owner, person1);
+      });
+      dog2.get('owner').then(function (owner) {
+        assert.equal(owner, person1);
+      });
+    });
+  });
+});
+
+// skip("saved changes to relationships should not roll back to a pre-saved state (from parent)", function(assert) {
```

Comment:
So I'll add my 2c:

> In order to rollback from the parent side it's not just as simple as reverting back to the relationship's canonical state... For example, there may have been unsaved objects added to that relationship. What would one do with those? Blindly delete them? What if they previously belonged to another parent? Do we blindly add them back to that other parents relationship? What if that other parent had been changed in the meantime and been saved, now what happens to that child?

So when I think of parent.rollback() in the above code you noted, I see it simply as, rollback changes that occurred on that("parent") record where changed have been only the relationship state between it and the child records. The unsaved objects that are children should be handled as a separate concern, where you reference them and roll them back should you want to. Or alternatively when a parent wants to rollback more than just relationship state, it should be a different method or an option passed to rollback ? `parent.rollback({cascade:true})` or `parent.rollbackChildren()`